taichi/transforms/whole_kernel_cse.cpp

```diff
@@ -67,9 +67,26 @@ class WholeKernelCSE : public BasicStmtVisitor {
     visited_.insert(stmt->instance_id);
   }
 
+  static std::size_t operand_type_hash(const Stmt *stmt) {
+    std::size_t hash_code{0};
+    auto hash_type =
+        std::hash<std::type_index>{}(std::type_index(typeid(stmt)));
+    auto op = stmt->get_operands();
+    for (auto &x : op) {
+      if (x == nullptr)
+        continue;
+      hash_code =
+          (hash_code) ^
+          (std::hash<unsigned long>{}(reinterpret_cast<unsigned long>(x)));
```

Comment:
Are you hashing the address of the pointer `x` here? If so, this might cause some missing optimization opportunities -- Note that `common_statement_eliminable` only returns `same_statements(this_stmt, prev_stmt)` when the statement's type is not `GlobalPtrStmt` or `LoopUniqueStmt`. In the `GlobalPtrStmt` and `LoopUniqueStmt` cases, even if 2 statements have different operands (but the same value), `common_statement_eliminable` can be `true`.

One suggestion is to simply not put the operands into the hash value when the statement type is `GlobalPtrStmt` or `LoopUniqueStmt`.

Also, only doing XOR is not a good hash function because we want to distinguish `[op1, op2]` and `[op2, op1]` cases. We can do a time33 hash like this:

```suggestion
      // Hash the addresses of the operand pointers.
      hash_code =
          (hash_code * 33) ^
          (std::hash<unsigned long>{}(reinterpret_cast<unsigned long>(x)));
```

Reply:
I do really hashing the address of pointer here. For example `GlobalPtrStmt`, we know that the position of `Stmt` which is the operand of `GlobalPtrStmt` is always before the position of this `GlobalPtrStmt`. If two `Stmt` operands have the same address, it will be CSEed formerly. Due to the `modified` parameter, CSE will re-run recursively.

Reply:
I also have a question: If CSE has worked correctly, are there existing two different pointers that have the same value?

Reply:
The short answer is I don't know. There can be two statements that have the same value but cannot be CSEd because the former is not visible to the latter, or one of them has side effects (e.g. global pointers with activate=true).

Even if CSE guarantees that there are no such cases, assuming it not necessarily makes the pass faster -- it may cause to run several more CSE passes. So I would suggest being conservative here to not miss any optimization chances. (If the user cares about the compilation time more, they can simply turn optimization_level down to disable this pass.)

Reply:
OK! I will be conservative to deal with it.